app/validators/questionnaire_schema.py

```diff
@@ -0,0 +1,298 @@
+import collections
+from collections import defaultdict
+from functools import cached_property, lru_cache
+import jsonpath_rw_ext as jp
+from jsonpath_rw import parse
+
+
+def has_default_route(routing_rules):
+    for rule in routing_rules:
+        if "goto" not in rule or "when" not in rule["goto"].keys():
+            return True
+    return False
+
+
+def get_routing_when_list(routing_rules):
+    when_list = []
+    for rule in routing_rules:
+        when_clause = rule.get("goto", {})
+        when_list.append(when_clause)
+    return when_list
+
+
+def is_contained_in_dict_list(dict_list, key_id):
+    for dict_to_check in dict_list:
+        if dict_to_check["id"] == key_id:
+            return True
+    return False
+
+
+def find_duplicates(values):
+    return [item for item, count in collections.Counter(values).items() if count > 1]
+
+
+@lru_cache
+def get_key_index_from_path(key, path):
+    position = path.find(key) + len(key + ".[")
+    return int(path[position : position + 1])
+
+
+@lru_cache
+def get_element_path(key, path):
+    position = path.find(key)
+    return path[: position + len(key + ".[0]")]
+
+
+# pylint: disable=too-many-public-methods
+class QuestionnaireSchema:
+    def __init__(self, schema):
+        self.schema = schema
+
+        self.blocks = jp.match("$..blocks[*]", self.schema)
+        self.sub_blocks = jp.match(
+            "$..[add_block, edit_block, add_or_edit_block, remove_block]", self.schema
+        )
+        self.blocks_by_id = {
+            block["id"]: block for block in self.blocks + self.sub_blocks
+        }
+        self.block_ids = list(self.blocks_by_id.keys())
+        self.sections = jp.match("$.sections[*]", self.schema)
+        self.sections_by_id = {section["id"]: section for section in self.sections}
+        self.section_ids = list(self.sections_by_id.keys())
+
+        self.groups = jp.match("$..groups[*]", self.schema)
+        self.groups_by_id = {group["id"]: group for group in self.groups}
+        self.group_ids = list(self.groups_by_id.keys())
+
+        self.list_names = jp.match(
+            '$..blocks[?(@.type=="ListCollector")].for_list', self.schema
+        )
+
+    @cached_property
+    def metadata_ids(self):
+        if "metadata" in self.schema:
+            return [m["name"] for m in self.schema["metadata"]]
+        return []
+
+    @cached_property
+    def questions_with_context(self):
+        for match in parse("$..question").find(self.schema):
+            yield match.value, self.get_context_from_path(str(match.full_path))
```

Comment:
`get_context_from_path` is passed the stringified JSONPath, but then does string manipulation to build the context. Can the match object be used for this in the same way that `match.context.value` is used further down this file?